sys/include/oorb.h

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_oorb    Thread-safe ringbuffer for one writer and one reader
+ * @ingroup     sys
+ * @brief       Thread-safe ringbuffer for one writer and one reader
+ *              implementation
+ * @{
+ *
+ * @note        Use @ref sys_tsrb instead, if you have more than one thread
+ *              writing to or more than one thread reading from the ring buffer
+ *
+ * @attention   Buffer size must be a power of two!
+ *
+ * @warning     This is only safe to use when (in addition to only one reader
+ *              and one writer being present) the target platforms implements
+ *              writes to `unsigned int` atomically (e.g. not on AVR) *and*
+ *              the compiler does not reorder code across memory accesses to
+ *              `volatile`. Improving this situation is planned.
+ *
+ * @file
+ * @brief       Thread-safe ringbuffer for one writer and one reader interface
+ *              definition
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef OORB_H
+#define OORB_H
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief     thread-safe ringbuffer struct
+ */
+typedef struct oorb {
+    uint8_t *buf;               /**< Buffer to operate on. */
+    unsigned int size;          /**< Size of buffer, must be power of 2. */
+    volatile unsigned reads;    /**< total number of reads */
+    volatile unsigned writes;   /**< total number of writes */
```

Comment:
Do those need to be `volatie`?`

Reply:
No, those should be atomic. Or more precisely, semi-atomic :-D

(As this is the two-threads-one-buffer mode, only the reader will write to `reads` and only the writer will write to `writes`. So as long as the read and the write access to them is atomically, everything is fine. This means explicitly that a read-modify-write operation could consist of three operations with only the read and write being atomic. Real atomic accesses would make the whole read-modify-write atomically. The atomic utils helper functions introduce just the semi-atomic functions this needs - which at least for word sized variables should be lock free.)

But I intentionally left the implementation of oorb unchanged. Fixing it efficiently would require the atomic utils PR being merged. And fixing it inefficiently is done in tsrb.